plugin/plugin.go

```diff
@@ -0,0 +1,210 @@
+// when the command is executed
+func (p *Plugin) SetStdIn(r io.Reader) *Plugin {
+	p.stdIn = r
+	return p
+}
+
+// SetStdOut configures to where the plugin writes to
+// when the command is executed
+func (p *Plugin) SetStdOut(w io.Writer) *Plugin {
+	p.stdOut = w
+	return p
+}
+
+// SetStdErr configures to where the plugin writes errors to
+// when the command is executed
+func (p *Plugin) SetStdErr(w io.Writer) *Plugin {
+	p.stdErr = w
+	return p
+}
+
+// SetEnv configures the available environment for the plugin
+func (p *Plugin) SetEnv(env []string) *Plugin {
+	p.env = env
+	return p
+}
+
+// Exec executes the plugin command
+func (p *Plugin) Exec(ctx context.Context, args []string) error {
+	// Prepare env so plugin has Dir available
+	p.setDirInEnv()
+	main, cmdArgs, err := p.MetaData.Command.Prepare()
+	if err != nil {
+		return fmt.Errorf("plugin exec prepare: %v", err)
+	}
+
+	execArgs := append(cmdArgs, args...)
+	cmd := exec.CommandContext(ctx, main, execArgs...)
+	cmd.Stdin = p.stdIn
+	cmd.Stdout = p.stdOut
+	cmd.Stderr = p.stdErr
+	cmd.Env = p.env
+
+	if err = cmd.Run(); err != nil {
+		// Check for a 1 exit status to check if it is a conftest test failure
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				if status.ExitStatus() == 1 {
+					return nil
+				}
+
+				return fmt.Errorf("plugin exec: %v", err)
+			}
+		}
+
+		return fmt.Errorf("plugin exec: %v", err)
+	}
+
+	return nil
+}
+
+func (p *Plugin) setDirInEnv() {
+	// Use os.SetEnv as the plugin needs access to this environment variable
+	os.Setenv("CONFTEST_PLUGIN_DIR", p.Dir)
+}
+
+// LoadPlugin loads the plugin.yaml from the given path
+// and parses the metadata into a Plugin struct
+func LoadPlugin(path string) (*Plugin, error) {
+	pluginPath := filepath.Join(path, metadataFileName)
+	data, err := ioutil.ReadFile(pluginPath)
+	if err != nil {
+		return nil, fmt.Errorf("read plugin file: %v", err)
+	}
+
+	plugin := &Plugin{
+		Dir:    path,
+		stdIn:  os.Stdin,
+		stdOut: os.Stdout,
+		stdErr: os.Stderr,
+		env:    os.Environ(),
+	}
+
+	if err := yaml.Unmarshal(data, &plugin.MetaData); err != nil {
+		return nil, fmt.Errorf("parse plugin file: %v", err)
+	}
+
+	return plugin, nil
+}
+
+// FindPlugins returns a list of all plugins available on the local file system.
+func FindPlugins() ([]*Plugin, error) {
+	var plugins []*Plugin
+	homePath, err := fetchHomeDir()
+	if err != nil {
+		return nil, fmt.Errorf("fetch home path: %v", err)
+	}
+
+	pluginsCacheDirPath := filepath.Join(homePath, ConftestDir, PluginsCacheDir)
+	if _, err := os.Stat(pluginsCacheDirPath); os.IsNotExist(err) {
+		// No plugins, so just return the empty slice
+		return plugins, nil
+	}
+
+	files, err := ioutil.ReadDir(pluginsCacheDirPath)
+	if err != nil {
+		return nil, fmt.Errorf("read plugin cache directory: %v", err)
+	}
+
+	// TODO log here instead? Otherwise conftest breaks on faulty plugins in the cache
```

Comment:
Imho, it should stay as it is (break when we face with any faulty on the plugin)